### PR TITLE
[SYCL] Emit supressed warnings

### PR DESCRIPTION
**Pull request by srividya-sundaram**
This PR updates FE tests related to the following warnings:

- Passing kernel functions by value is deprecated in SYCL 2020.
- Passing of kernel functions by reference is a SYCL 2020 extension.
- SYCL 1.2.1 specification requires an explicit forward declaration for a kernel type name; your program may not be portable.

The updated tests include the mock <sycl.hpp> header file and call the kernel invocation functions defined in the mock header.

### CODE REVIEW

File: clang/test/CodeGenSYCL/Inputs/sycl.hpp [hunk 1]
```diff
@@ -281,6 +281,11 @@ ATTR_SYCL_KERNEL void kernel_single_task(const KernelType &kernelFunc) {
   kernelFunc();
 }
 
+template <typename KernelName = auto_name, typename KernelType>
+ATTR_SYCL_KERNEL void kernel_single_task_2017(KernelType kernelFunc) {
```
Comment (srividya-sundaram) [inline clang/test/CodeGenSYCL/Inputs/sycl.hpp:285]: This was added to differentiate SYCL 2017 kernel invocation function from SYCL 2020.
The SYCL 2020 version supported passing kernel functions by reference, while SYCL 2017 version supported passing kernel function by value.

File: clang/test/CodeGenSYCL/kernel-by-reference.cpp [hunk 2]
```diff
@@ -1,38 +1,33 @@
-// RUN: %clang_cc1 -triple spir64 -fsycl -fsycl-is-device -verify -fsyntax-only -sycl-std=2017 -DSYCL2017 %s
-// RUN: %clang_cc1 -triple spir64 -fsycl -fsycl-is-device -verify -fsyntax-only -sycl-std=2020 -DSYCL2020 %s
-// RUN: %clang_cc1 -triple spir64 -fsycl -fsycl-is-device -verify -fsyntax-only -Wno-sycl-strict -DNODIAG %s
-// RUN: %clang_cc1 -triple spir64 -fsycl -fsycl-is-device -verify -fsyntax-only -sycl-std=2020 -Wno-sycl-strict -DNODIAG %s
+// RUN: %clang_cc1 -triple spir64 -fsycl -fsycl-is-device -internal-isystem %S/Inputs -verify -fsyntax-only -sycl-std=2017 -DSYCL2017 %s
+// RUN: %clang_cc1 -triple spir64 -fsycl -fsycl-is-device -internal-isystem %S/Inputs -verify -fsyntax-only -sycl-std=2020 -DSYCL2020 %s
+// RUN: %clang_cc1 -triple spir64 -fsycl -fsycl-is-device -internal-isystem %S/Inputs -verify -fsyntax-only -Wno-sycl-strict -DNODIAG %s
+// RUN: %clang_cc1 -triple spir64 -fsycl -fsycl-is-device -internal-isystem %S/Inputs -verify -fsyntax-only -sycl-std=2020 -Wno-sycl-strict -DNODIAG %s
```
Comment (srividya-sundaram) [inline clang/test/CodeGenSYCL/kernel-by-reference.cpp:4]: I am not sure what value the RUN command at L4 adds. `-Wno-sycl-strict` flag disables all warnings in the sycl-strict diag group(adding the -sycl-std=2020 doesn't seem to make a differnence) and we already have a similar check at L3. I can remove L4 if not needed.

File: clang/test/SemaSYCL/args-size-overflow.cpp [hunk 3]
```diff
@@ -1,36 +1,25 @@
-// RUN: %clang_cc1 -fsycl -triple spir64 -fsycl-is-device -fsyntax-only -verify %s
-// RUN: %clang_cc1 -fsycl -triple spir64 -Werror=sycl-strict -DERROR -fsycl-is-device -fsyntax-only -verify %s
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -internal-isystem %S/Inputs -fsyntax-only -Wsycl-strict -sycl-std=2020 -verify %s
```
Comment (srividya-sundaram) [inline clang/test/SemaSYCL/args-size-overflow.cpp:1]: Added `-sycl-std=2020` flag so we don't get this warning : `Passing of kernel functions by reference is a SYCL 2020 extension` , since the original test is about warning related to size of kernel arguments.

File: clang/test/SemaSYCL/implicit_kernel_type.cpp [hunk 4]
```diff
@@ -1,47 +1,16 @@
...
-#else
-    kernelFunc();
-#endif
-  }
-};
-// -- /Definitions
-
-#ifdef __SYCL_UNNAMED_LAMBDA__
-// expected-no-diagnostics
-#endif
```
Comment (srividya-sundaram) [inline clang/test/SemaSYCL/implicit_kernel_type.cpp:38]: This condition was no longer true. With the updated test, passing `-fsycl-unnamed-lambda` caused the following error: Passing of kernel functions by reference is a SYCL 2020 extension. I added an #if check to capture that.

File: clang/test/SemaSYCL/args-size-overflow.cpp (hunk 3 above)
Comment (srividya-sundaram) [reply clang/test/SemaSYCL/args-size-overflow.cpp:1]: Also, removed this test from `sycl/test/basic_tests/args-size-overflow.cpp`

File: clang/test/SemaSYCL/args-size-overflow.cpp [hunk 5]
```diff
@@ -1,36 +1,25 @@
...
-    cl::sycl::accessor<int, 1, cl::sycl::access::mode::read_write, cl::sycl::access::target::global_buffer>;
+int main() {
 
-void use() {
   struct S {
     int A;
     int B;
-    Accessor AAcc;
-    Accessor BAcc;
-    int Array[1991];
```
Comment (Fznamznon) [inline clang/test/SemaSYCL/args-size-overflow.cpp:31]: Why did you remove them?
The test in sycl/basic_tests was simplified and needed only to make sure that the *warning is emitted*. This one should remain *as is* but also test that warning is emitted for system headers.

File: clang/test/SemaSYCL/args-size-overflow.cpp [hunk 6]
```diff
@@ -1,36 +1,25 @@
...
+    int Array[2015];
   } Args;
-  auto L = [=]() { (void)Args; };
-  // expected-note@+1 {{in instantiation of function template specialization 'parallel_for<Foo}}
-  parallel_for<Foo>(L);
```
Comment (Fznamznon) [inline clang/test/SemaSYCL/args-size-overflow.cpp:35]: And why did you remove it?

File: clang/test/CodeGenSYCL/kernel-by-reference.cpp [hunk 7]
```diff
@@ -1,38 +1,33 @@
...
 
-// SYCL 1.2/2017 - kernel functions passed directly. (Also no const requirement, though mutable lambdas never supported)
-template <typename name, typename Func>
-#if defined(SYCL2020)
-// expected-warning@+2 {{Passing kernel functions by value is deprecated in SYCL 2020}}
-#endif
-__attribute__((sycl_kernel)) void sycl_2017_single_task(Func kernelFunc) {
-  kernelFunc();
-}
+#include "Inputs/sycl.hpp"
```
Comment (Fznamznon) [inline clang/test/CodeGenSYCL/kernel-by-reference.cpp:6]: Including mock header through full patch disregard `internal-isystem` option, so we do not emulate sycl header being in system headers
```suggestion
#include "sycl.hpp"
```

File: clang/test/SemaSYCL/implicit_kernel_type.cpp [hunk 8]
```diff
@@ -1,47 +1,16 @@
...
+#include "Inputs/sycl.hpp"
```
Comment (Fznamznon) [inline clang/test/SemaSYCL/implicit_kernel_type.cpp:5]: Same thing here.
```suggestion
#include "sycl.hpp"
```

File: sycl/test/basic_tests/args-size-overflow.cpp [hunk 9]
```diff
@@ -1,21 +0,0 @@
-// RUN: %clangxx -fsycl -fsycl-device-only -Xclang -verify -Wsycl-strict %s -fsyntax-only
```
Comment (s-kanaev) [inline sycl/test/basic_tests/args-size-overflow.cpp:1]: Why was this test deleted?

Comment (Fznamznon) [reply sycl/test/basic_tests/args-size-overflow.cpp:1]: This test was added to make sure that the warning is emitted for SYCL headers which are included as system headers. It turned out that it is possible to emulate such include scheme in FE tests as well, so we decided to remove this test to avoid duplication, because we are going to change FE tests to use "system headers include scheme".

**Review by s-kanaev — APPROVED** [review]
RT tests changes LGTM.

File: clang/test/SemaSYCL/implicit_kernel_type.cpp [hunk 10]
```diff
@@ -50,34 +19,70 @@ class myWrapper2;
...
+#if defined(WARN)
+  // expected-warning@Inputs/sycl.hpp:211 {{Passing of kernel functions by reference is a SYCL 2020 extension}}
+  // expected-error@Inputs/sycl.hpp:220 {{kernel needs to have a globally-visible name}}
+  // expected-note@+8 {{InvalidKernelName1 declared here}}
+#elif defined(ERROR)
+  // expected-error@Inputs/sycl.hpp:211 {{Passing of kernel functions by reference is a SYCL 2020 extension}}
+  // expected-error@Inputs/sycl.hpp:220 {{kernel needs to have a globally-visible name}}
+  // expected-note@+4 {{InvalidKernelName1 declared here}}
+#elif defined(__SYCL_UNNAMED_LAMBDA__)
+  // expected-error@Inputs/sycl.hpp:211 {{Passing of kernel functions by reference is a SYCL 2020 extension}}
```
Comment (Fznamznon) [inline clang/test/SemaSYCL/implicit_kernel_type.cpp:32]: Since the test is not about the difference between ways of passing kernel functions in different sycl standards, can we suppress this error/warning by passing `-sycl-std=2020` like you did in the other test?

**Review by Fznamznon — COMMENTED** [review]
One small nit, otherwise LGTM

**Review by Fznamznon — APPROVED** [review]
